Make the /json endpoint return all data

trailtacos opened the pull request:
This change alters /json to return all the sensor data (as it
would post to a telemetry collector), and changes the main
UI index page to solely use data from /json to create the page,
including the initial view. This moves things like the AQI CSS
color styles into the Javascript, and removes all the variables
populated in index.html

This both simplifies the C++ code, and enables more advanced data
collection from things like grafana/telegraf

Commit:
Make the /json endpoint return all data

This change alters /json to return all the sensor data (as it
would post to a telemetry collector), and changes the main
UI index page to solely use data from /json to create the page,
including the initial view. This moves things like the AQI CSS
color styles into the Javascript, and removes all the variables
populated in index.html

This both simplifies the C++ code, and enables more advanced data
collection from things like grafana/telegraf

trailtacos commented:
@luejerry I'm tagging you here - as this PR changes your JSON endpoint to use the JSON data that is emitted to the telemetry collector - and moves the color calculation into the HTML - in fact it moves all dynamic data into the HTML/JS rather than templating via the C++ code.

trailtacos commented:
A future PR will document how to use this /json endpoint with grafana/telegraf/influxdb - which is my actual goal here.

michaelkamprath commented on data/script.js:
```diff
@@ -39,22 +31,36 @@ function showAQI(evt, avr_window) {
   evt.currentTarget.classList.add("active");
 }
 
+function aqiToColorClass(val) {
+  if (val <= 50) {
+    return 'aqi-green';
+  } else if (val <= 100) {
+    return 'aqi-yellow';
+  } else if (val <= 150) {
+    return 'aqi-orange';
+  } else if (val <= 200) {
+    return 'aqi-red';
+  } else if (val <= 300) {
+    return 'aqi-purple';
+  } else {
+    return 'aqi-maroon';
+  }
+}
+
```
My moving this into the JavaScript, the system now has two parallel implementations of the AQI-to-color mapping. You can't get rid of the C++ version because it is needed for things other than HTML rendering, and the javascript version can't satisfy C++ needs. Why not simply use the C++ version? What's the thinking of having this also in javascript?

trailtacos replied:
There are really two parallel implementations in the C++ anyway - they just shared a common enum - the goal here was to remove the CSS class names from the C++ code - having the backend tell the front-end specific CSS classes to use breaks the separation of front-end/backend.

One alternative is to alter the enum in the C++ to be something like "AQI_HEALTHY", "AQI_MODERATE" etc - and then expose "healthy" / "moderate" through the json - and then allow the front-end to pick CSS classes based on those names (which then aren't describing specific colors, but more the state) - then the LED mapping would just use the new enum for its color code...

Note also - I wanted to have one JSON data format for both the "push" and "pull" models of getting data - hence I needed to alter the previous /json endpoint which was subtly different from the json payload being "pushed" - I think that is a good design approach to use (single json payload for everything)

michaelkamprath replied:
I get the one json format thing. No critique there.

I do think it is better for the AQI value to band transformation be on the C++ side, so putting the band enumeration into the JSON makes sense. But I would argue per the EPA definition of AQI bands, the color is synonymous with "healthy", "moderate", etc. There is no need for the ["healthy", "moderate",...] abstraction, just puts the color name in the JSON.  That is, just return a text representation of what `AirQualitySensor::getAQIStatusColor()` returns, and then have your switch statement in javascript to turn that into a CSS class name.

The C++ LED color function then continue to use the same enum then.

trailtacos replied:
Would you prefer the JSON be altered to go back to how @luejerry had it (with `value` and `color` values), or kept backwards compatible with the color values added back as separate entries in the `air_quality_index` object, e.g.

```
	"air_quality_index": {
		"aqi_current": 0,
                "aqi_current_color": "green"
        }
```
vs
```
	"air_quality_index": {
		"aqi_current": {
                        "value": 0,
                        "color": "green"
        }
```

michaelkamprath replied:
> Would you prefer the JSON be altered to go back to how @luejerry had it 

Honestly, your call. I am chiefly concerned with not duplicating the logic of what AQI maps to what color.

trailtacos replied:
Are you currently using the data from the "push" JSON message?  I like @luejerry 's format - but don't want to break backwards compatibility (I can probably work around any change with my grafana install)

michaelkamprath replied:
Correct me if I am wrong, but it looks like you are just adding elements. That won't break anything on my end. Even if it did, it's not like this is a "production system". I'd rather push forward rather than be beholden to a hacked first implementation.

michaelkamprath reviewed (commented):
Generally looks good, but I do have one question here.

Commit:
Ensure all color calculation is in the C++

This change alters the JSON response for both the push and pull
handlers (its consistent) to pass the color to the front-end.
The front-end then converts that to the appropriate CSS style
(which for AQI is simply a matter of prefixing it with `aqi-`)

michaelkamprath reviewed (approved):
LGTM
